File: static/viewer/gemeinsam/lagenmaske.js
/**
 * Lagenmaske — welche Punkte eines Kleidungsstücks unter einem ANDEREN Stück
 * liegen.
 *
 * BEFUND (11.09.2026, Dance1, T-Shirt über Leggings): Der Bund der Leggings
 * drückte in Bewegung durch das T-Shirt — schwarze Flecken im roten Stoff.
 * Dieselbe Sache wie die Haut unter dem Stoff (`hautmaske.js`): zwei
 * getrennt gehäutete Flächen, wenige Millimeter auseinander, kommen sich an
 * jedem Gelenk nahe. Und dieselbe Lösung: Was unter einem anderen Stück
 * liegt, wird nicht gezeichnet.
 *
 * WER LIEGT AUSSEN, wird gemessen, nicht angenommen: Je Paar zählt, wie
 * viele Punkte von A ein Stück B mindestens `MINDEST_M` vor sich haben, und
 * umgekehrt; die Mehrheit entscheidet. Bei Hose über T-Shirt (Bund über den
 * Saum gebaut, 11.09.2026) gewinnt die Hose, bei T-Shirt über Leggings das
 * T-Shirt — die Reihenfolge des Anziehens sagt darüber nichts, und beim
 * Laden einer Szene gibt es sie gar nicht.
 *
 * DIE NORMALEN KOMMEN VON DER HAUT, nicht vom Stoffnetz: Jeder Stoffpunkt
 * bekommt die Normale des nächsten Körperpunkts (wie `hautmitstoff.py` auf
 * dem Server). Die Dreiecksorientierung eines Schnitts kann zum Körper
 * zeigen — das Vorzeichen daraus hat auf dem Server viermal in Folge das
 * Falsche gemeldet (CLAUDE.md, 07.–08.09.2026). Die Haut ist geschlossen,
 * ihr Vorzeichen steht fest.
 *
 * Die Maske selbst rechnet `Hautmaske.verdeckt` mit dem inneren Stück als
 * „Körper": Strahl entlang der Hautnormale, bis 25 mm darüber, 5 mm
 * Toleranz nach innen, freier Streifen nur an lockeren Kanten des äußeren
 * Stücks. OHNE THREE.JS — prüfbar in Node (`test_js_lagenmaske.py`).
 */
import { Hautmaske } from './hautmaske.js';
import { Hautmaskegeometrie as G } from './hautmaskegeometrie.js';

export class Lagenmaske {

    /** So weit muss das andere Stück mindestens VOR einem Punkt liegen,
     *  damit es für die Lagenfrage als „außen" zählt. */
    static MINDEST_M = 0.001;

    /**
     * Je Stück: `{maske, ueber}` — `maske` je Punkt 1, wenn ein anderes Stück
     * darüber liegt; `ueber` die Schlüssel dieser Stücke.
     *
     * @param koerper  punkte und dreiecke — die Haut in Ruhelage
     * @param stoffe   Liste von Stücken (schluessel, punkte, dreiecke) in derselben Lage
     * @param optionen wie bei `Hautmaske.verdeckt` (abstand, tiefe, randringe, eng)
     * @return Map schluessel → {maske, ueber}
     */
    static verdeckt(koerper, stoffe, optionen = {}) {
        const N = G.normalen(koerper.punkte, koerper.dreiecke);
        const gitter = G.punktgitter(koerper.punkte, G.ZELLE_M);
        const normalen = stoffe.map((s) => Lagenmaske.normalenVonHaut(s.punkte, koerper.punkte, N, gitter));
        const ergebnis = new Map(stoffe.map((s) => [
            s.schluessel, { maske: new Uint8Array(s.punkte.length / 3), ueber: [] }]));
        for (let a = 0; a < stoffe.length; a++) {
            for (let b = a + 1; b < stoffe.length; b++) {
                const lage = Lagenmaske.lage(stoffe[a], normalen[a], stoffe[b], normalen[b], optionen);
                if (lage === 0) continue;
                const [innen, aussen, nInnen] = lage > 0
                    ? [stoffe[a], stoffe[b], normalen[a]] : [stoffe[b], stoffe[a], normalen[b]];
                const m = Hautmaske.verdeckt(innen.punkte, innen.dreiecke, [aussen], { ...optionen, normalen: nInnen });
                const eintrag = ergebnis.get(innen.schluessel);
                for (let i = 0; i < m.length; i++) eintrag.maske[i] |= m[i];
                eintrag.ueber.push(aussen.schluessel);
            }
        }
        return ergebnis;
    }

    /**
     * +1: B liegt über A; −1: A liegt über B; 0: die beiden berühren sich
     * nicht. Gezählt werden Punkte, die das andere Stück mindestens
     * `MINDEST_M` vor sich haben (binnen `abstand`).
     */
    static lage(A, nA, B, nB, optionen = {}) {
        const o = { ...optionen, tiefe: -Lagenmaske.MINDEST_M, randringe: 0, inseln: 0,
                    suchweite: optionen.abstand ?? Hautmaske.ABSTAND_M };
        const bUeberA = Lagenmaske._summe(Hautmaske.verdeckt(A.punkte, null, [B], { ...o, normalen: nA }));
        const aUeberB = Lagenmaske._summe(Hautmaske.verdeckt(B.punkte, null, [A], { ...o, normalen: nB }));
        if (!bUeberA && !aUeberB) return 0;
        return bUeberA >= aUeberB ? 1 : -1;
    }

    /** Je Stoffpunkt die Normale des nächsten Körperpunkts — (0,0,0), wenn
     *  in drei Zellen (9 cm) keiner liegt; ein Nullstrahl trifft nichts. */
    static normalenVonHaut(P, koerperP, N, gitter) {
        const aus = new Float64Array(P.length);
        for (let i = 0; i < P.length / 3; i++) {
            const nah = G.naechsterPunkt(koerperP, gitter, G.ZELLE_M, P[3 * i], P[3 * i + 1], P[3 * i + 2]);
            if (!nah) continue;
            aus[3 * i] = N[3 * nah.j]; aus[3 * i + 1] = N[3 * nah.j + 1]; aus[3 * i + 2] = N[3 * nah.j + 2];
        }
        return aus;
    }

    static _summe(maske) {
        let s = 0;
        for (let i = 0; i < maske.length; i++) s += maske[i];
        return s;
    }
}

File: TheatreJS/ModelPhysik/lagenmaske.py
# -*- coding: utf-8 -*-
u"""Stoff unter Stoff: welche Punkte eines Stuecks unter einem anderen liegen.

Herausgeloest aus `hautmaske.py` (12.09.2026). Wer aussen liegt, wird je Paar
GEZAEHLT (Punkte mit dem anderen Stueck mindestens `MINDEST_M` vor sich); die
Normalen kommen von der Haut, nie vom Stoffnetz — dessen Dreiecksorientierung
kann zum Koerper zeigen. Browser-Fassung: `gemeinsam/lagenmaske.js`.
"""
import numpy as np
from scipy.spatial import cKDTree

from hautmaske import Hautmaske
from maskengeometrie import Geometrie


class Lagenmaske:
    u"""Welche Punkte eines Stuecks unter einem anderen liegen."""

    MINDEST_M = 0.001

    @classmethod
    def verdeckt(cls, koerper, koerperdreiecke, stoffe, **optionen):
        u"""`stoffe`: Liste von (name, punkte, dreiecke). Liefert
        {name: (maske, [namen der Stuecke darueber])}."""
        koerper = np.asarray(koerper, dtype=np.float64)
        N = Geometrie.normalen(koerper, koerperdreiecke)
        baum = cKDTree(koerper)
        normalen = [cls.normalen_von_haut(np.asarray(p, dtype=np.float64), N, baum)
                    for _n, p, _t in stoffe]
        aus = {name: (np.zeros(len(p), dtype=bool), [])
               for name, p, _t in stoffe}
        for a in range(len(stoffe)):
            for b in range(a + 1, len(stoffe)):
                lage = cls.lage(stoffe[a], normalen[a], stoffe[b], normalen[b],
                                **optionen)
                if lage == 0:
                    continue
                innen, aussen, n_innen = ((stoffe[a], stoffe[b], normalen[a])
                                          if lage > 0 else
                                          (stoffe[b], stoffe[a], normalen[b]))
                m = Hautmaske.verdeckt(innen[1], innen[2], [(aussen[1], aussen[2])],
                                       normalen=n_innen, **optionen)
                aus[innen[0]][0][:] |= m
                aus[innen[0]][1].append(aussen[0])
        return aus

    @classmethod
    def lage(cls, A, nA, B, nB, **optionen):
        o = dict(optionen)
        o.update(tiefe=-cls.MINDEST_M, randringe=0, inseln=0,
                 suchweite=optionen.get('abstand', Hautmaske.ABSTAND_M))
        b_ueber_a = int(Hautmaske.verdeckt(A[1], None, [(B[1], B[2])],
                                           normalen=nA, **o).sum())
        a_ueber_b = int(Hautmaske.verdeckt(B[1], None, [(A[1], A[2])],
                                           normalen=nB, **o).sum())
        if not b_ueber_a and not a_ueber_b:
            return 0
        return 1 if b_ueber_a >= a_ueber_b else -1

    @staticmethod
    def normalen_von_haut(P, N, baum):
        d, i = baum.query(P, distance_upper_bound=0.09, workers=-1)
        aus = np.zeros_like(P)
        nah = np.isfinite(d)
        aus[nah] = N[i[nah]]
        return aus
